### PR TITLE
WIP: Vector parameters

### PR DESCRIPTION
**Pull request by TTitscher**
Solving #8 
I basically added a member latent._index_length that keeps track of the length of the parameters. 

a -> 1
v -> [1, 2, 3, 4, 5]

latent.add("a" ...)
latent.add("v" ...)

latent._index_length --> [1, 5]

In latent.update(numbers) the numbers (in this case of required length 6) are split into two vectors (for two parameters) of length 1 and 5 and assigned accordingly.

Could we live with something like that?

[Still WIP as following things like defining priors for those are not yet considered.]

### CODE REVIEW

**Comment by joergfunger** [comment]
That looks already very good. I think it would be nice to have an additional test case that illustrates and tests this vector parameter in a joint list (so that this vector values parameter is a latent parameter in a joint parameter list).

File: tests/test_inference_problem.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        return time_steps, sensor_data
+
+    data1 = generate_data(101, noise_sd = 0.1)
+    data2 = generate_data(51, noise_sd = 0.2)
+
+    me1 = MyModelError(fw, data1)
+    me2 = MyModelError(fw, data2)
+
+    problem = InferenceProblem()
+    key1 = problem.add_model_error(me1, fw.parameter_list())
```
Comment (joergfunger) [inline tests/test_inference_problem.py:354]: I would suggest that the model error generates this list, not the forward model. This allows to retrieve in the model_error the parameter_list, but potentially modify values that are already given, e.g. the measured environmental temperature that we do not want to touch in the inference, or e.g. the transverse forces in Matthis compression tests that have been measured (but we don't want to infer). This model error could even by used with the same parameterized forward model to compute e.g. the tensile tests with/without transverse expansion. And here is another question - not directly related to the MR, but do we distinguish inputs to the forward model that are static (should/could be set when creating the model and are currently directly given in the constructor - not via a parameter_list but one by one) and those that are dynamic (stored in the parameter_list).

**Review by joergfunger — COMMENTED** [review]
Great work, I know its WIP, so maybe some of the comments are already planned to be resolved.

File: tests/test_inference_problem.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+    problem.latent.add("A", "A", key1)
+    problem.latent.add("A", "A", key2)
+    problem.latent.add_by_name("B")
+
+    problem.latent.set_prior("A", NormalDistribution(40, 5))
+    problem.latent.set_prior("B", NormalDistribution(6000, 300))
+
+    tmp = problem([A_correct, B_correct])
+
+    problem.noise.add("exp1_noise", s1, key1)
```
Comment (joergfunger) [inline tests/test_inference_problem.py:366]: I think for the naming, I would prefer sensor_group, since this is what the function does, grouping sensormodel errors into groups. The specific noise is just an addon, where you specify for each noise group a corresponding noise term in VB. So for me, there you create a problem.sensor_group.add("force_sensors", s1,key), and only afterwards relate a specific sigma to this sensor group problem.noise_group.add("force_sensors", sigma_force).

File: tests/test_inference_problem.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    print(info)
+
+    print(1.0 / info.noise.mean[0] ** 0.5)
+    print(1.0 / info.noise.mean[1] ** 0.5)
+
+    pymc3_problem = PyMC3Problem(problem)
+    print(pymc3_problem([0, 1, 2, 3]))
+
+    def max_likelihood(numbers):
+        return -pymc3_problem(numbers)
```
Comment (joergfunger) [inline tests/test_inference_problem.py:396]: for clarity, I think the call function in pyMC3 should not be the loglike, I would prefer to name the function as loglike.

File: tests/test_inference_problem.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+    result = scipy.optimize.minimize(max_likelihood, x0=[A_correct, B_correct, noise_sd1, noise_sd2], method="Nelder-Mead")
```
Comment (joergfunger) [inline tests/test_inference_problem.py:398]: the usual maximum-likelihood does not change the hyperparameters, so noise_sd1 would be given. Otherwise I would guess you always get the sd to approach infinity, because that minimizes the model error and thus maximizes the likelihood.

**Review by joergfunger — COMMENTED** [review]
I somehow missed some files, added more comments.

File: tests/test_inference_problem.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+class ModelError:
+    def __call__(self, parameter_list):
+        raise NotImplementedError("Override this!")
+
+
+class LatentParameter:
+    def __init__(self):
+        self._affected_parameters = []
+        self.N = None
+        self.prior = None
```
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py:49]: Indeed.
* For the _superuser_, we could return (or provide access) to the internal ordering and he can do whatever he wants
* Your second idea could also work out of the box, but is then tool specific.
~~~py
with py.model():
   inference_problem.latent.set_prior("my_parameter, pm.pymc3.distributions.continuous.Kumaraswamy(6174., 42.)
~~~

File: tests/test_inference_problem.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        return f"{idx_range:10} {self._affected_parameters} {self.prior}"
+
+
+class LatentParameterList:
+    def __init__(self):
+        self._all_parameter_lists = {}
+        self._latent = OrderedDict()
+        self._total_length = None
+
+    def add_parameter_list(self, parameter_list, key=None):
```
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py:81]: Good point. But 'model_parameter_list' would be more suited as I came across the case where a parameter was part of the model_error -- e.g. related to scaling of the data -- but not the forward model.

File: tests/test_inference_problem.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joergfunger) [reply tests/test_inference_problem.py:49]: IMO, the question is what we want to achieve. On the one hand, we would like to provide an interface for the optimizer to the model/model error. On the other hand, optimization/inference in itself could be tool independent, but I would say this is less important. If the second option works, that would be fully sufficient.

File: tests/test_inference_problem.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (joergfunger) [reply tests/test_inference_problem.py:81]: Then we should maybe change that to model error prms, or live with the fact that for specific cases that fw_model_prm is not a fully correct naming. In the Bayesian community, they often also talk about modelling, but they mean the uncertainty model, so I think it's important to avoid model as a single term.

File: tests/test_inference_problem.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+                self._all_parameter_lists[key][prm_name] = latent_numbers
+
+    def __str__(self):
+        s = ""
+        for key, latent in self._latent.items():
+            s += f"{key:10}: {latent} \n"
+        return s
+
+
+class Noise:
```
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py]: We have to think about that and I was just toying around. I separated it because "LatentParameterList" (currently) only works on model parameters. 

I could not really come up with a nice way to include it..
~~~py
problem.noise.add("noise_name", sensor5, exp3_key)
# and later
problem.latent.add("noise_name", ???) 
~~~
Issue 1: They are _only_ coupled by the "noise_name" which the user has to drag around, if he wants to automate that. 
Issue 2: What to put in "???" because there is no ParameterList associated to the noise.

So one idea could be to have a 
~~~py
noise_prm = problem.noise.get_parameter_list()

noise_key = "SomeReallyDistinctKeyThatDoesNotInterfereWithPossibleUserKeys"
problem.latent.add_parameter_list(noise_prm, key=noise_key)
~~~

... actually quite OKish/goodish ;)

File: tests/test_inference_problem.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    def __init__(self):
+        self._noise_groups = OrderedDict()
```
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py]: The whole question of correlation is not yet tackled and I assume that will change things again.

So far, even the pyMC3_Wrapper is not implemented. I imagine that the "loglikelihood_function" could be overwritten, if needed.

File: tests/test_inference_problem.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+###############################################################################
+"""
+
+class MySensor(Sensor):
+    def __init__(self, name, position):
+        super().__init__(name, shape=(1,1))
+        self.position = position
+
+class MyForwardModel:
+    def __call__(self,parameter_list, sensors, time_steps):
```
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py]: In our demotwin, it made sense to separate the time steps. Conceptually, the sensor is really "just there" (i.e. attached somewhere) and can measure stuff. The measurement system then associates the sensor values with time information. Translated to our library, the model itself depends on the time information (e.g. time dependent boundary conditions) and the sensor, again, just measures. 

Practically, I can use the same sensor in multiple measurements where the time steps (I actually meant `time_stamps` or `times`, but that does not matter much) differ.

File: tests/test_inference_problem.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+        p = ParameterList()
+        p.define("A", None)
+        p.define("B", None)
+        return p
+
+
+class MyModelError(ModelError):
+    def __init__(self, fw, data):
+        self._fw = fw
+        self._ts, self._sensor_data = data
```
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py:301]: It is! `model_response = self._fw(prm, sensors, self._ts)`

File: tests/test_inference_problem.py (hunk 1 above)
Comment (TTitscher) [reply tests/test_inference_problem.py:354]: First part: The parameter list generation/creation is not part of the interface. That means that there is no automatic treatment of the paramter lists, e.g. 
~~~py
for model_error in inference_problem._model_errors.values():
    prm = model_error.define_parameter_list()
    # now what?
~~~
That is because the user _has_ to touch this list to be forced to set values explicitly. Thus, it is his choice when/how to define it. So I agree, it can make sense for the user to implement his model_error such that it defines this list or implement it like I did before like:
~~~py
prm1 = fw.define_parameter_list() + model_error1.define_additional_parameters()
prm2 = fw.define_parameter_list() + model_error2.define_additional_parameters()
~~~

Long story short: It should not be forced when/how it is created.

Second part: I would not put too much effort into that because ultimately it is the users problem how to define his model/model_error.

File: tests/test_inference_problem.py [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+        return self._all_parameter_lists
+
+    def __str__(self):
+        s = ""
+        for key, latent in self.items():
+            s += f"{key:10}: {latent} \n"
+        return s
+
+
+class NoiseGroup(list):
```
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py:130]: I should adapt the naming. 
NoiseGroup: list of (key, sensor) to keep track of the individual contributions to the noise group

NoiseGroups: "named" list == dict of multiple NoiseGroup

**Comment by TTitscher** [comment]
Thanks for your feedback, almost everything good and "solvable" IMO. 

And I am sorry for accidentally mixing up the branches, but I'll now leave it here because of our discussion.

@joergfunger If you follow the existing structure of the main file, how would you like to include information on how to correlate what? I mean the maths is (for easy cases) clear, but what would be a convenient/reasonable/readable way to tell the library what to do?

File: tests/test_inference_problem.py (hunk 2 above)
Comment (TTitscher) [reply tests/test_inference_problem.py:366]: Following the idea of a "noise parameter list", I can imagine the following:
~~~py
problem.sensor_group.add("force_sensors", ...)

noise_prm = problem.sensor_group.define_parameters()

# now either:
noise_prm["force_sensors"] = sigma_force
# or
problem.latent.add_parameter_list(noise_prm, noise_key)
problem.latent.add_by_name("force_sensors")
problem.latent.set_prior("force_sensors", Gamma(...))
~~~

Comment (TTitscher) [reply tests/test_inference_problem.py:366]: Note: this hardcodes exactly one parameter for each sensor group which may not be sufficient...

File: tests/test_inference_problem.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py:81]: Well, I just had a look again. In this specific case "parameter_list" could be fine, because noise_parameter_lists or hyper_parameter_list could also be stored there.

File: tests/test_inference_problem.py [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+
+    problem.latent.set_prior("A", NormalDistribution(40, 5))
+    problem.latent.set_prior("B", NormalDistribution(6000, 300))
+
+    tmp = problem([A_correct, B_correct])
+
+    noise_model1 = UncorrelatedNoise()
+    noise_model2 = UncorrelatedNoise()
+    noise_key1 = problem.sensor_groups.define("exp1_noise", noise_model1)
+    noise_key2 = problem.sensor_groups.define("exp2_noise", noise_model2)
```
Comment (TTitscher) [inline tests/test_inference_problem.py:385]: Note to self.
A cleaner solution would be to almost 100% follow the way the model errors are treated. So it would read 
~~~
noise_key1 = problem.add_noise_model(noise_model1, noise_model1.define_parameter_list())
~~~
which automatically creates a sensor group (now identified by its key(!!!) instead of some name) and puts the noise parameter list to the (potentially) latent parameters.

The whole key generation should be revisited as it is essential for VB (and also helpful for the user) to distinguish between `model_error_parameters` and `noise_model_parameters`.

File: tests/test_inference_problem.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (joergfunger) [reply tests/test_inference_problem.py]: I would have a 
```
problem.inference_prms.add('sigma_1', 0.1)
problem.inference_prms.add('YoungsModulus', stats.normal(30.000, 10.000))
```
the first variable is deterministic with a given value, the other is latent (due to giving a prior). These latent variables are the ones we loop over for the prior. As for the independent model errors I would suggest something like
```
inference_prms['sd1'] = 0.1 #deterministic
inference_prms['sd2'] = 10 #deterministic
inference_prms['sd3'] = stats.Gamma(s=0.1, c=3) #latent not sure about the actual distribution
inference_prms['l_corr_time] = 0.5 #correlation length
inference_prms['l_corr_space] = 0.5 

independent_model_errors.add_log_like(IndependentZeroMeanNormalNoiseModelError(key,sensor, 'sd1'))
independent_model_errors.add_log_like(TimeCorrelatedZeroMeanNormalNoiseModelError(key, sensor, time_steps, 'sd2', 'l_corr_time') #len(timesteps)=len(model_error(key,sensor)
independent_model_errors.add_log_like(SpaceTimeCorrelatedZeroMeanNormalNoiseModelError(key, sensor, time_steps, coordinates, 'sd3', 'l_corr_time', 'l_corr_space) #we will have to define the structure of a space_time_model error
```
And someone that wants to correlated two different sensortypes in space (e.g. because there is a laser that measures some points correlated in space and a stereo camera that measures also correlated in space, but both sensors are clearly not independent since they measure the same thing)
```
independent_model_errors.add_log_like(MyFancyModelError(key1, sensor1, key2, sensor2, coordinates1, coordinates2, 'sd1', 'sd2', 'l_corr_space', 'l_corr_laser_stereo')
```
 so in the likelihood, there will be a loop over the indenpent_model_errors each returning the contribution to the loglikelihood (a scalar value). In the special case of VB, you would only consider model errors of type IndependentZeroMeanNormalNoiseModelError (naming is certainly to be discussed)

File: tests/test_inference_problem.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (joergfunger) [reply tests/test_inference_problem.py]: That is true, but image the case of having two different sensors with different measurement frequencies. What are you returning in the output for each individual sensor, since this should not be the union of all sensor time steps. If so, you would have to manually decompose the results, since each sensor type has only data for a subset of all time steps.

**Comment by joergfunger** [comment]
Maybe we should step back and try to think how the user would implement his/her problem.
1. The user defines a (set of) forward models (e.g. compression test, tensile test) that has as input a set of parameters, and as output something related to each individual sensor. In the current approach, this is a scalar/vector. Maybe this should already include spatial/temporal information, so that we that each sensoroutput is e.g. of type Vector, TimeDependentVector, SpaceDependentVector or SpaceTimeDependentVector (this should cover everything). The forward model also provides a method that creates a fw_prms (initialized with Nones for values supposed to be added by the inference)

1. The user defines a model error that reads in (either from a file or from the database..) and internally stores the data as well as modifies the fw_prms in order to include metadata (e.g. in the metadata of the compression test this could be the information of the measured environmental temperature, the size of the specimen, the time steps with data). Note that it's the users choice if he/she adds the information directly to the constructor (e.g. in case of a mesh this seems a reasonable choice) or if it's added as a parameter that can be changed. This choice is primarily influenced by the fact, whether we could potentially think about making these parameters latent. From a personal point of view, I think it should be easy to move parameters from constructor to the fw_prms and vice versa, I would prefer a dict type of approach for both. (and it simplifies the documentation since both dicts can just be serialized).

1. The user defines an InferenceProblem. Then a loop over all experiments to be added is performed
e.g. each 
```
inference_problem = InferenceProblem() 
inference_problem.add_inference_prm('E', someDistributionWithParameters) #note that the distribution parameters given can refer to other parameters previously defined here 
inference_problem.add_inference_prm('sigma_f', 0.1) #deterministic 
inference_problem.add_inference_prm('l_corr', 2.) #deterministic 
for test in tensile_tests_yaml_files:  #or whatever is the loop that provides the data
    # define model error
    ################
    model_error = tensile_test_model_error(my_fw_tensile_model, ..)
    fw_prm = model_error.get_parameter()
    key_this_tensile_test = inference_problem.add_me(model_error, fw_prm)
    
    # define fw_prm and map to inference_prm 
    ################
    # this is for shared variables - we need to map the inference prm to the fw model parameters 
    inference_problem.map_inference2fw_prm('E', key_this_tensile_test, 'Youngs modulus')
    # for variables that are individual for each experiment, we can simultaneously add them to the global list 
    # inference_prm with a given key are automatically local
    inference_problem.add_inference_prm('offset', key_this_tensile_test, 'initialOffset')
    
    # define groups of model errors that each contribute similarly to the loglike i.e. noise groups 
    inference_problem.loglike.add_me(key_model_error, 'T', 'sigma_T') # just uncorrelated sensors
    inference_problem.loglike.add_me_correlated_in_time(key_this_tensile_test, 'force_sensor', 'sigma_f', kernelfunction('l_corr'))    
```
In the final application we then implement a loglike function that just loops over the loglike.me and sums up the individual contributions.

File: tests/test_inference_problem.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py]: When it comes to e.g. model selection, I think it is advantageous for the user to automatically set deterministic parameter from some parameter files. So they should not explicitly be mentioned in the main file. The idea would be that we ask the model "what parameters to you need" and then look up a parameter file (or set them explicitly if needed) to provide those. The latent parameters, however, are in full control of the user and must be explicitly mentioned. And I would stick to this pattern for the noise models as well. So similar to the model_errors, the noise_models can tell the parameter it needs (initialized with `None`) and the user has to provide them. 

Your proposed idea is the other way round. You first define a parameter, e.g. "sd2" that you then pass to a noise model. The drawbacks I see:
* You accidentally write `independent_model_errors.add_log_like(TimeCorrelatedZeroMeanNormalNoiseModelError(key, sensor, time_steps, 'l_corr_time', 'sd2')` instead of your line. IMO hard to spot bug that "sd2" is now the fifth arg instead of fourth. That would also be against he "named parameter" idea because at this point, the meaning of the parameter is passed as the position in the argument list, not its name.
* I have to think about the focus (so mainly the naming) on "log_like". The user should IMO provide more general information that is later -- depending on the inference algorithm -- transformed into a loglike or a least-square vector. This information would just be: 
    * The output sensor2 in key1 is correlated in time with length 0.42s and has a zero mean noise with "sigma"=latent. 
    * The output of sensor1 in key1 is uncorrelated and has a systematic noise of a*t + b where a and b are latent.
    * The output of sensor3 in all keys (=model_errors) is correlated with 0.42s has known N(0., .6174)

The last example of combining individual coordinates is something we agreed not to allow. That came up when:
* one physical sensor has two noise terms (on bridge and off bridge). Here we agreed on splitting the sensor into two, each with his separate noise
* in a displacement field some displacements are already fixed via boundary conditions. So the "HugeDisplacementSensor" would be split accordingly.
* Same in your example: The measurement position of the stereo camera would also have to be split into individual sensors. 

But we may change that, sure.

File: tests/test_inference_problem.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (TTitscher) [reply tests/test_inference_problem.py]: Maybe we can skip part of the discussion here, as a forward model interface is not part of the library. We start at the "model_error". 

sensor1: 1Hz
sensor2: 100Hz

If I chose to evaluate that within the same forward_model or model_error, both would return (measured over 1 seconds) a dict like:
`{sensor1: <vector length 1>, sensor2: <vector length 100>}`

I do not understand where a union or decomposition is needed.

**Comment by TTitscher** [comment]
In short, I would agree with the steps. I would, however, treat the noise models similar to the model errors. 
1) The user creates it. 
2) The user adds it to the inference model which assigns both the noise model and the noise parameter list to a noise key. 
3) The noise parameter list is treated like the model error parameter list, so the user sets parameters, sets parameters latent, shares them, defines priors...
4) The treatment/processing of the information stays algorithm dependent.
